src/main/java/net/minecraftforge/common/attachment/AttachmentHolder.java

```diff
@@ -0,0 +1,121 @@
+package net.minecraftforge.common.attachment;
+
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraftforge.registries.ForgeRegistries;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Implementation class for objects that can hold data attachments.
+ * For the user-facing methods, see {@link IAttachmentHolder}.
+ */
+public abstract class AttachmentHolder<H extends AttachmentHolder<H>> implements IAttachmentHolder
+{
+    final Map<AttachmentType<?>, Object> attachments = new IdentityHashMap<>();
+
+    // TODO: should these methods validate that the attachment type is registered?
+    @Override
+    public final boolean hasData(AttachmentType<?> type)
+    {
+        Objects.requireNonNull(type);
+        return attachments.containsKey(type);
+    }
+
+    @Override
+    public final <T> T getData(AttachmentType<T> type)
+    {
+        Objects.requireNonNull(type);
+        return (T) attachments.computeIfAbsent(type, t -> Objects.requireNonNull(t.defaultValueSupplier.get()));
+    }
+
+    @Override
+    public final <T> void setData(AttachmentType<T> type, T data)
+    {
+        Objects.requireNonNull(type);
+        Objects.requireNonNull(data);
+        attachments.put(type, data);
+    }
+
+    @Nullable
+    public final CompoundTag serializeAttachments()
+    {
+        var reg = ForgeRegistries.ATTACHMENT_TYPES.get();
+
+        CompoundTag tag = null;
+        for (var entry : attachments.entrySet())
+        {
+            var type = entry.getKey();
+            if (type.serializer != null)
+            {
+                if (tag == null)
+                    tag = new CompoundTag();
+                tag.put(reg.getKey(type).toString(), ((IAttachmentSerializer<?, Object>) type.serializer).write(entry.getValue()));
+            }
+        }
+        return tag;
+    }
+
+    protected final void deserializeAttachments(CompoundTag tag)
+    {
+        var reg = ForgeRegistries.ATTACHMENT_TYPES.get();
+
+        for (var key : tag.getAllKeys())
+        {
+            var type = reg.getValue(new ResourceLocation(key));
+            if (type != null && type.serializer != null)
+            {
+                attachments.put(type, ((IAttachmentSerializer<Tag, ?>) type.serializer).read(tag.get(key)));
+            }
+        }
+    }
+
+    public final boolean areAttachmentsCompatible(H other)
```

Comment:
I know in one of the early concepts for caps, the idea of being able to provide a equality-comparing function was also mentioned, which I want to point out here. Comparison via serialization == could still be a default but is that something being considered (seeing as attachments already have a builder style API for implementing what you want here?)

Reply:
Definitely in scope, yes. I might leave it as "future work" however... ;)